### PR TITLE
#162725267 Enable users to add tags to their articles

### PR DESCRIPTION
**Pull request by kdaglas**
#### What does this PR do?
This PR enables a user to add tags to his articles and view all the available tags

#### Description of Task to be completed?
- When a user has created an article, it is always good culture to give a hint what that article is all about, most especially if the article is long and someone cannot read all of it or when there are too many stories and a user wants to find stories related to a particular thing. this is where tags come in.
- should have the endpoint `/api/tags` working
- create a tag field that will handle the data from the user as tags
- create a view function that fetches all tags

#### How should this be manually tested?
- Fetch this branch by running the command `git fetch origin ft-tag-articles-162725267` in the terminal and check into the branch `git checkout ft-tag-articles-162725267`
- Start the virtual environment `source venv/bin/activate`
- Migrate so that the fields required for the data are initiated by typing this command: `python manage.py migrate`
- Start the application by running this command: `python manage.py runserver`
- Go to postman and in it create an article with a tag 

#### Screenshots
`POST  /api/articles`
screen shot 2019-01-28 at 11 58 45 am

screen shot 2019-01-28 at 11 58 58 am

`GET  /api/tags`
screen shot 2019-01-28 at 12 07 53 pm

`PATCH /api/article/<article-slug>`
screen shot 2019-01-28 at 11 53 58 am

screen shot 2019-01-28 at 11 54 12 am

#### Related story
[#162725267](https://www.pivotaltracker.com/story/show/162725267)

[Run in Postman](https://app.getpostman.com/run-collection/9a1fb51a44517a13f838)

### CODE REVIEW

**Comment by MuhweziDeo** [comment]
@kdaglas  the implementation is on point especially adding the array field to hold tags other than creating a separate model for tags. Everything checks out when I ran the feature locally and the tests Maybe add your Postman documentation to your PR

**Review by MuhweziDeo — DISMISSED** [review]

**Comment by ja-odur** [comment]
@kdaglas Please add postman collections, it makes testing you PR much easier.

**Review by calvinpete — COMMENTED** [review]
@kdaglas please correct the git command for switching branches in your PR description testing section, secondly, we are pushing migrations which mean that you should update the testing section not to include making migrations.

**Comment by ja-odur** [comment]
@kdaglas please use `tags` instead of `all_tags`

File: authors/apps/articles/urls.py [hunk 1]
```diff
@@ -7,5 +7,7 @@
     path('<slug>/', views.ArticleUpdateDeleteView.as_view(), name='article-update'),
     path('<slug>/likes/', views.ArticleLikesView.as_view(), name='article-likes'),
     path('<slug>/bookmark', views.BookmarkAPIView.as_view(), name='article-bookmark'),
-    path('<slug>/rate/', views.RatingsAPIView.as_view(), name='article-rates')
+    path('<slug>/rate/', views.RatingsAPIView.as_view(), name='article-rates'),
+    # path('slug/tags', views.TagListAPIView.as_view(), name='article-tag'),
```
Comment (ja-odur) [inline authors/apps/articles/urls.py]: @kdaglas Please remove this commented code

File: authors/apps/articles/tests/test_tags.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    """
+
+    def test_fetch_all_tags_succeeds_if_authorized(self):
+        """
+        Testing if there is successful retrieval of all tags
+        """
+        response = self.client.get(reverse('tags'),
+                        HTTP_AUTHORIZATION='Bearer ' +
+                        self.test_user_token)
+        print(response.data)
```
Comment (ja-odur) [inline authors/apps/articles/tests/test_tags.py]: @kdaglas Please remove this print statement

**Comment by mariamiah** [comment]
screen shot 2019-01-28 at 11 12 17 am

@kdaglas i think you do not need the requirements command since you did not add any new requirements. Please remove it

**Comment by ja-odur** [comment]
@kdaglas please rebase and fix the dropping coverage

**Comment by ja-odur** [comment]
@kdaglas the `{{base-url}}/api/tags` is not protected, could there be any reason for this?

**Comment by kdaglas** [comment]
@ja-odur we wanted to make it in a way that if a user who is not logged in wants to check out stories, he can access them using tags then when he or she is to view that story, they need to be logged in

File: authors/apps/authentication/tests_/test_base_class.py [hunk 3]
```diff
@@ -111,7 +111,8 @@ def setUp(self):
         self.article = {
             "title": "hello worlfd",
             "description": "desctriptuo",
-            "body": 'boddydydabagd'
+            "body": 'boddydydabagd',
+            "all_tags": ["andela", "xmen"]
```
Comment (ja-odur) [inline authors/apps/authentication/tests_/test_base_class.py]: @kdaglas this has to be `tags` not `all_tags`.

Comment (kdaglas) [reply authors/apps/authentication/tests_/test_base_class.py]: oh yes, let me change that

**Comment by kdaglas** [comment]
@ja-odur I've protected the endpoint since earlier on we had protected the viewing all articles endpoint

**Review by ja-odur — APPROVED** [review]
LGTM

**Review by MuhweziDeo — APPROVED** [review]
All set

**Review by calvinpete — APPROVED** [review]
LGTM

**Review by d-rita — APPROVED** [review]
Works fine

**Pull request merged**